_posts/2018-10-23-artsy-auth.md

```diff
@@ -0,0 +1,269 @@
+access to this service, or part of it, or something like that. Maybe you only want Artsy employees to have access.
+Maybe you only want to allow requests from other artsy applications. Regardless, you need to figure out how to
+provision access to the functionality you've deployed. At Artsy, we use OAuth for that.
+
+OAuth is a paradigm that allows one service to delegate authentication to another one. You've probably used this
+before, even if you don't realize it. Whenever you use your Twitter account to log in to Medium, or your Facebook
+account to log in to Tinder, or whatever, you're relying on delegated authentication. Medium doesn't have to know
+much about who you are as long as they can trust Twitter to tell them, etc.
+
+We use a similar approach here at Artsy. We've got dozens of applications, some outward-facing, some internal only.
+If every single application had to handle its own authentication then we'd have user objects stored all over the
+place and we'd be constantly trying to figure out if User 123 on service A is the same person as User 321 on
+service B, etc, right? So instead we delegate authentication to a single core service, Gravity. Gravity is the
+canonical source of truth for user information, and that includes authentication logic.
+
+So at a high level, OAuth works like this:
+
+1. An HTTP request is sent to your server.
+2. Your server checks for the presence of an `Authorization` HTTP header.
+3. If it cannot find one, your server redirects the request to login through Gravity.
+4. If the user logs in through Gravity, Gravity gathers a bunch of information about that user, encrypts it, and
+   gives it to the user to give your server by adding the encrypted string to the user's `Authorization` header.
+5. The user is then redirected back to your server, this time with the header in place. Your server tries to
+   decrypt the header, using a private key that it shares with Gravity.
+6. If that private key works and the header is decrypted, your server now has access to whatever information about
+   the user that gravity shared.
+
+There's a slight difference in authenticating a _user_ and authenticating a _trusted application_. If our incoming
+request was generated by another Artsy service then we don't redirect anyone to gravity for login - rather, we
+manually create the token in advance, and store it. Once this is set up, the above process exits after step 2. But
+for now, let's continue looking at how user authentication works, because it's important to understand the whole
+system. We'll circle back to app authentication at the end.
+
+So!
+
+Think about it this way: when the user is redirected to gravity for login, gravity gathers up a bunch of
+information about who this user is. What's their user ID? What roles has their account been assigned? It lumps
+these together to create a JWT, which is a short statement of identity verification. It sticks this JWT into the
```

Comment:
We could link to JWT.io or http://artsy.github.io/blog/2016/10/26/jwt-artsy-journey/ at some point.